File: binarytree/binarytree.go
package main

import (
	"log"
	"fmt"
)

type Node struct {
	value int
	rightchild *Node
	leftchild *Node
	parent *Node
}

type BinaryTree struct {
	root *Node
}

func ( bt *BinaryTree ) makeNode( paramValue int ) *Node {
	return &Node{ paramValue, nil, nil, nil }
}

func ( bt *BinaryTree ) insert( paramValue int ) (bool, error) {
	if bt.root == nil {
		bt.root = bt.makeNode( paramValue )
		return true, nil
	} else {
		parent := bt.root
		node := bt.root
		for node != nil {
			if node.value >= paramValue {
				parent = node
				node = node.leftchild
			} else {
				parent = node
				node = node.rightchild
			}
		}
		node = bt.makeNode( paramValue )
		if node.value <= parent.value {
			parent.leftchild = node
			node.parent = parent
			return true, nil
		} else {
			parent.rightchild = node
			node.parent = parent
			return true, nil
		}
	}
	return false, fmt.Errorf("Failed to insert: %v", paramValue)
}

func ( bt *BinaryTree ) remove( paramValue int ) (bool, error) {
	parent := bt.root
	node := bt.root
	for node != nil {
		if node.value > paramValue {
			parent = node
			node = node.leftchild
		} else if node.value < paramValue {
			parent = node
			node = node.rightchild
		} else {
			if node == parent.leftchild {
				if node.leftchild == nil && node.rightchild == nil {
					parent.leftchild = nil
					return true, nil
				} else if node.leftchild != nil && node.rightchild == nil {
					parent.leftchild = node.leftchild
					return true, nil
				} else if node.leftchild == nil && node.rightchild != nil {
					parent.leftchild = node.rightchild
					return true, nil
				} else {
					temp, err := bt.minNode(node.rightchild)
					parent.leftchild.value = temp.value
					if temp == temp.parent.leftchild {
						temp.parent.leftchild = nil
					} else {
						temp.parent.rightchild = nil
					}
					return true, err
				}
			} else {
				if node.leftchild == nil && node.rightchild == nil {
					parent.rightchild = nil
					return true, nil
				} else if node.leftchild != nil && node.rightchild == nil {
					parent.rightchild = node.leftchild
					return true, nil
				} else if node.leftchild == nil && node.rightchild != nil {
					parent.rightchild = node.rightchild
					return true, nil
				} else {
					temp, err := bt.minNode(node.rightchild)
					parent.rightchild.value = temp.value
					if temp == temp.parent.leftchild {
						temp.parent.leftchild = nil
					} else {
						temp.parent.rightchild = nil
					}
					return true, err
				}
			}
		}
	}
	return false, fmt.Errorf("Failed. Value does not exist: %v", paramValue)
}

func ( bt *BinaryTree ) minNode( paramRoot *Node ) (*Node, error) {
	if paramRoot == nil {
		return nil, fmt.Errorf("Failed. Nil tree you fool.")
	}
	result := paramRoot
	for result.leftchild != nil {
		result = result.leftchild
	}
	return result, nil
}

func ( bt *BinaryTree ) has( paramValue int ) bool {
	node := bt.root
	for node != nil {
		if node.value > paramValue {
			node = node.leftchild
		} else if node.value < paramValue {
			node = node.rightchild
		} else {
			return true
		}
	}
	return false
}

func main() {
	var binaryTree BinaryTree

	for i := 0 ; i < 10 ; i++ {
		_, ok := binaryTree.insert(i)
		if ok != nil {
			log.Println(ok)
		}
		for j := 20 ; j >10 ; j-- {
			_, ok = binaryTree.insert(j)
			if ok != nil {
				log.Println(ok)
			}
		}
	}

	for i := 0 ; i < 10 ; i++ {
		_, ok := binaryTree.remove(i)
		if ok != nil {
			log.Println(ok)
		}
		/*
		for j := 20 ; j > 10 ; j-- {
			_, ok = binaryTree.remove(j)
			if ok != nil {
				log.Println(ok)
			}
		}
		*/
	}

	for k := 0 ; k <= 20 ; k++ {
		log.Println(binaryTree.has(k))
	}
}

File: binarytree/binarytree.js
function Node() {
	"use strict";
	var value = null;
	var leftchild = null;
	var rightchild = null;
	var parent = null;
}

function BinaryTree() {
	"use strict";
	this.root = null;

	this.makeNode = function() {
		var node = new Node();
		return node;
	};
	this.insert = function( paramValue ) {
		if ( !this.root ) {
			this.root = this.makeNode();
			this.root.value = paramValue;
			return true;
		} else {
			var parent = this.root;
			var node = this.root;
			while ( node ) {
				if ( node.value >= paramValue ) {
					parent = node;
					node = node.leftchild;
				} else {
					parent = node;
					node = node.rightchild;
				}
			}
			node = this.makeNode();
			node.value = paramValue;
			if ( node.value <= parent.value ) {
				parent.leftchild = node;
				node.parent = parent;
				return true;
			} else {
				parent.rightchild = node;
				node.parent = parent;
				return true;
			}
		}
		return false;
	};

	this.remove = function( paramValue ) {
		var parent = this.root;
		var node = this.root;
		while ( node ) {
			if ( node.value > paramValue ) {
				parent = node;
				node = node.leftchild;
			} else if (node.value < paramValue) {
				parent = node;
				node = node.rightchild;
			} else {
				if ( node === parent.leftchild ) {
					if ( !node.leftchild && !node.rightchild ) {
						delete parent.leftchild;
						return true;
					} else if ( node.leftchild && !node.rightchild ) {
						parent.leftchild = node.leftchild;
						return true;
					} else if ( !node.leftchild && node.rightchild ) {
						parent.leftchild = node.rightchild;
						return true;
					} else {
						var temp = this.minNode(node.rightchild);
						parent.leftchild.value = temp.value;
						if ( temp === temp.parent.leftchild ) {
							temp.parent.leftchild = null;
						} else {
							temp.parent.rightchild = null;
						}
						return true;
					}
				} else {
					if ( !node.leftchild && !node.rightchild ) {
						delete parent.rightchild;
						return true;
					} else if ( node.leftchild && !node.rightchild ) {
						parent.rightchild = node.leftchild;
						return true;
					} else if ( !node.leftchild && node.rightchild ) {
						parent.rightchild = node.rightchild;
						return true;
					} else {
						var temp = this.minNode(node.rightchild);
						parent.rightchild.value = temp.value;
						if ( temp === temp.parent.leftchild ) {
							parent.leftchild = null;
						} else {
							parent.rightchild = null;
						}
						return true;
					}
				}
			}
		}
		return false;
	};

	this.minNode = function ( paramRoot ) {
		var result = paramRoot;
		while ( result.leftchild ) {
			result = result.leftchild;
		}
		return result;
	};

	this.has = function( paramValue ) {
		var node = this.root;
		while ( node ) {
			if ( node.value > paramValue ) {
				node = node.leftchild;
			} else if (node.value < paramValue) {
				node = node.rightchild;
			} else {
				return true;
			}
		}
		return false;
	};

}



